packages/zoe/src/contractFacet/zcfZygote.js

```diff
@@ -364,48 +359,52 @@ export const makeZCFZygote = (
    * @type {ZCFZygote}
    * */
   const zcfZygote = {
-    evaluateContract: bundleOrBundleCap => {
-      let bundle;
-      if (passStyleOf(bundleOrBundleCap) === 'remotable') {
-        const bundleCap = bundleOrBundleCap;
-        // @ts-expect-error vatPowers is not typed correctly: https://github.com/Agoric/agoric-sdk/issues/3239
-        bundle = powers.D(bundleCap).getBundle();
-      } else {
-        bundle = bundleOrBundleCap;
-      }
-      contractCode = evalContractBundle(bundle);
-      handlePWarning(contractCode);
-    },
-    startContract: (
+    // wire zcf up to zoe instance-specific interfaces
+    startContract: async (
       instanceAdminFromZoe,
       instanceRecordFromZoe,
       issuerStorageFromZoe,
       privateArgs = undefined,
     ) => {
       zoeInstanceAdminPromiseKit.resolve(instanceAdminFromZoe);
+      zcfBaggage.init('instanceAdmin', instanceAdminFromZoe);
       instantiateInstanceRecordStorage(instanceRecordFromZoe);
       instantiateIssuerStorage(issuerStorageFromZoe);
 
-      // Next, execute the contract code, passing in zcf
-      /** @type {Promise<ExecuteContractResult>} */
-      const result = E(contractCode)
-        .start(zcf, privateArgs)
-        .then(
-          ({
-            creatorFacet = Far('emptyCreatorFacet', {}),
-            publicFacet = Far('emptyPublicFacet', {}),
-            creatorInvitation = undefined,
-          }) => {
-            return harden({
-              creatorFacet,
-              publicFacet,
-              creatorInvitation,
-              handleOfferObj,
-            });
-          },
-        );
-      handlePWarning(result);
-      return result;
+      // If it's a non-upgradeable contract, start it now.
+      if (!setupInstallation) {
+        return startClassicContract(privateArgs);
+      }
+
+      // now that our clone is differentiated, we can do
+      // instance-specific setup and get back the contract runner
+      const makeInstanceKit = await setupInstance(privateArgs);
+
+      // and invoke makeInstanceKit() for the first and only time
+      const { publicFacet, creatorFacet } = makeInstanceKit();
+      /** @type {ExecuteUpgradeableContractResult} */
+      const upgradeableResult = harden({
+        publicFacet,
+        creatorFacet,
+        handleOfferObj,
+      });
+      return upgradeableResult;
+    },
+    restartContract: async (privateArgs = undefined) => {
```

Comment:
given `startContract` it looks like this must never be called until after a successful `startContract`. Can runtime assertions here confirm that?